File: python/transition_animation.py
from cycle import Cycle


class TransitionAnimation:
    def __init__(self,
                 transition_duration_ticks,
                 panel_context):
        self.panel_context = panel_context
        self.cycle = Cycle(transition_duration_ticks,
                           lambda value: self.on_up_cycle(value),
                           None,
                           False)
        self.current_value = 0
        self.target_gradient = None
        self.active = False

    def on_up_cycle(self, value):
        self.current_value = value

    def start(self,
              target_gradient):
        self.target_gradient = target_gradient
        self.cycle.restart()
        self.active = True

    def update(self):
        if not self.active:
            return

        self.cycle.next()

        if self.current_value >= .97:
            self.panel_context.switch_to_next_gradient()
            # Done!
            self.cycle.stop()
            self.active = False
            return

    def get_color(self, panel_value):
        current_gradient = self.panel_context.current_gradient
        from_color = current_gradient.get_color_for_value(panel_value)
        to_color = self.target_gradient.get_color_for_value(panel_value)

        return from_color.interpolate(to_color, self.current_value)

File: arduino/Aunisoma-Sketch/TransitionAnimation.cpp
//
// Created by David Clay on 6/16/23.
//

#include "TransitionAnimation.h"

#define ONE_SHOT true

TransitionAnimation::TransitionAnimation(int duration) {
    this->cycle = new Cycle(duration, ONE_SHOT, UP_ONLY_CYCLE);
    this->current_value = 0;
    this->active = false;
}


void TransitionAnimation::start(GradientValueMap* current_gradient, GradientValueMap* next_gradient) {
    this->current_gradient = current_gradient;
    this->next_gradient = next_gradient;
    // TODO: if this was already started, this will jump...
    // Note: restart() will reset the `iterations` so we can
    // determine whether the animation has run once vs not run yet.
    this->cycle->restart();
    this->active = true;
}

void TransitionAnimation::update() {
    if (!this->active) {
        return;
    }

    this->cycle->update();

    if (this->cycle->isDone()) {
        this->cycle->stop();
        this->active = false;
    }
}

bool TransitionAnimation::is_done() {
    // how do I know if it's done? It has to have _run_ to be done.
    return this->cycle->isDone();
}

/**
 * Used to reset the state of the transition animation so it can
 * be started again, vs knowing it is done and has stopped.
 */
void TransitionAnimation::reset() {
    this->cycle->reset();
}

Color TransitionAnimation::get_color(float panel_value) {
    Color from_color = this->current_gradient->getColorForValue(panel_value);
    Color to_color = this->next_gradient->getColorForValue(panel_value);

    Color color = from_color.interpolate(to_color, this->cycle->current_value);

    return color;
}
